api/src/models/FedoraObject.ts

```diff
@@ -145,6 +145,21 @@ export class FedoraObject {
         await this.addDatastreamFromStringOrBuffer(licenseXml, stream, "text/xml", [201, 204]);
     }
 
+    async modifyAgents(stream: string, agents): Promise<void> {
+        const agentsXml = agents.reduce((acc,agent)=> {
+            const { role, type, name, notes }= agent;
+            const notesXml = notes.reduce((acc,note) => {
+                return acc + `<METS:note>${xmlescape(note)}</METS:note>`
+            }, "");
+            const agentXml = `<METS:agent ROLE="${xmlescape(role)}" TYPE="${xmlescape(type)}"><METS:name>${name}</METS:name>${notesXml.trim()}</METS:agent>`.trim();
+            return acc + agentXml;
+        }, "");
+        const documentXml = `
+            <METS:metsHdr xmlns:METS="http://www.loc.gov/METS/" CREATEDATE="${new Date().toISOString()}" LASTMODDATE="${new Date().toISOString()}" RECORDSTATUS="">${agentsXml}</METS:metsHdr>
```

Comment:
This is now going to change the create date every time the record is modified. We need to establish the create date when the datastream is first created, and then never change it when other parts of the stream are modified. I think the simplest approach may be to actually pass the original create date as part of the post when the agents are modified. Obviously this would lend itself to spoofing/manipulation (as opposed to reading the value from the existing datastream before rewriting it, which is another possible solution) but it seems simpler and faster, and I can't think of a reason why anyone would want to falsify this data, or why it would matter.

Reply:
I decided to remove the security issue, by keeping the extraction of the create date in the backend.